source/common/http/http2/conn_pool.h

```diff
@@ -81,19 +94,63 @@ class ConnPoolImpl : public ConnectionPool::Instance, public ConnPoolImplBase {
 
   virtual CodecClientPtr createCodecClient(Upstream::Host::CreateConnectionData& data) PURE;
   virtual uint32_t maxTotalStreams() PURE;
+  /*
   void movePrimaryClientToDraining();
+  */
   void onConnectionEvent(ActiveClient& client, Network::ConnectionEvent event);
   void onConnectTimeout(ActiveClient& client);
+  void onIdleTimeout(ActiveClient& client);
   void onGoAway(ActiveClient& client);
   void onStreamDestroy(ActiveClient& client);
   void onStreamReset(ActiveClient& client, Http::StreamResetReason reason);
-  void newClientStream(Http::StreamDecoder& response_decoder, ConnectionPool::Callbacks& callbacks);
-  void onUpstreamReady();
+  void attachRequestToClient(ActiveClient& client, Http::StreamDecoder& response_decoder,
+                             ConnectionPool::Callbacks& callbacks);
+  void createNewConnection();
+  void onUpstreamReady(ActiveClient& client);
+  ConnectionPool::Cancellable* newStream(Http::StreamDecoder& response_decoder,
+                                         ConnectionPool::Callbacks& callbacks);
 
+  void applyToEachClient(std::list<ActiveClientPtr>& client_list,
+                         const std::function<void(const ActiveClientPtr&)>& fn);
   Stats::TimespanPtr conn_connect_ms_;
   Event::Dispatcher& dispatcher_;
-  ActiveClientPtr primary_client_;
-  ActiveClientPtr draining_client_;
+
+  // Connecting clients list. Connections remain in the list till either:
+  //   - Connection is established
+  //   - Connection time out
+  //   - Connection terminated (by remote)
+  std::list<ActiveClientPtr> connecting_clients_;
+
+  // Connected clients waiting for requests. Connections remain in this list
+  // till:
+  //  - Request is attached
+  //  - Connection terminated
+  std::list<ActiveClientPtr> ready_clients_;
+
+  // Connections serving at least one request. Connections remain in this list
+  // till:
+  //  - Max requests per connection is exceeded
+  //  - Connection terminated
+  std::list<ActiveClientPtr> busy_clients_;
+
+  // Connections with number of requests >= max requests per connection. Connections
+  // remain in this list till:
+  //  - Excess requests in the connections finish.
+  //  - Connection terminated
+  std::list<ActiveClientPtr> overflow_clients_;
+
+  // Connections that have active requests but can no longer accept new
+  // requests. Depending on the connection policy, these connections could be moved to
+  // `to_close_clients` list. Connections could also be moved to this list upon:
+  //    - timeout.
+  //    - if requests exceed the max allowed by connection policy.
+  std::list<ActiveClientPtr> drain_clients_;
+
+  // Connections that are waiting to be closed. Connections are moved to this
+  // list when drain clients does not have any more requests being served.
+  // Connections remain in this list till:
+  //  - Connection is closed.
```

Comment:
Why is this state needed? Can't we just close them directly vs. putting them on a new list?

Reply:
The idea is that a `busy/active` client could be transitioned to `drained` if the policy dictates a) to not accept any more requests on that client b) close the client once all current requests are finished.

Reply:
Sorry I don't follow. Once a client can be closed why does it need to be on a list?